### DISCUSSION TITLE
Issue with adding the GRPC library to the Freeplane plugin

### DISCUSSION THREAD

**Discussion by metacoma** [freeplane/freeplane#950] [Help] [answered]
Hey folks!
 
I am not very familiar with Java programming and the Java build system. I am trying to write a Freeplane plugin that exports the Freeplane Java API as a gRPC network service. 

Here is my code: https://github.com/metacoma/freeplane_plugin_grpc

I have completed the gRPC-Java tutorial, but when I move the gRPC code to the Freeplane project, I am getting a strange error when I run the Freeplane, I encountered the following exception:

```
Exception in thread "AWT-EventQueue-0" java.lang.NoClassDefFoundError: io/grpc/BindableService
        at org.freeplane.plugin.grpc.Activator$1.installExtension(Activator.java:21)
        at org.freeplane.main.osgi.ActivatorImpl$OsgiExtentionInstaller.installExtensions(ActivatorImpl.java:256)
        at org.freeplane.main.osgi.ActivatorImpl.installControllerExtensions(ActivatorImpl.java:272)
        at org.freeplane.main.osgi.ActivatorImpl.access$100(ActivatorImpl.java:62)
        at org.freeplane.main.osgi.ActivatorImpl$1.run(ActivatorImpl.java:196)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.ClassNotFoundException: io.grpc.BindableService
        at org.knopflerfish.framework.BundleClassLoader.findClass(BundleClassLoader.java:176)
        at org.knopflerfish.framework.BundleClassLoader.loadClass(BundleClassLoader.java:305)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 18 more
```

I searched on Google for this problem and found some "strange" solutions like the one on this link: https://stackoverflow.com/questions/42540531/bindableservice-issue-with-grpc-java

Could anyone please shed some light on this issue or help me with it?

**Comment by metacoma** [comment, marked as answer]
i don't know why, but this helps:

https://github.com/metacoma/freeplane_plugin_grpc/commit/9f250f4f4fc88a454b85900733caf4f00a653e6d# 

TLDR, i use `lib` keyword instead of `implementation` in `build.gradle` file

Now, after running a Freeplane instance with this plugin, Freeplane is able to interact with GRPC clients using the GRPC protocol.

```
$ gradle build
$ BIN/freeplane.sh &
$ netstat -nlp 2>&1 | grep 50051
tcp6       0      0 :::50051                :::*                    LISTEN      104921/java
$ ~/grpc-java/build/install/examples/bin/hello-world-client
Jan 18, 2023 4:32:39 PM io.grpc.examples.helloworld.HelloWorldClient greet
INFO: Will try to greet world ...
Jan 18, 2023 4:32:40 PM io.grpc.examples.helloworld.HelloWorldClient greet
INFO: Greeting: Hello world
```
